espnet/nets/pytorch_backend/e2e_asr_transducer.py

```diff
@@ -0,0 +1,414 @@
+        xs = [xx[::self.subsample[0], :] for xx in xs]
+        xs = [to_device(self, to_torch_tensor(xx).float()) for xx in xs]
+        xs_pad = pad_list(xs, 0.0)
+
+        # 0. Frontend
+        if self.frontend is not None:
+            enhanced, hlens, mask = self.frontend(xs_pad, ilens)
+            hs_pad, hlens = self.feature_transform(enhanced, hlens)
+        else:
+            hs_pad, hlens = xs_pad, ilens
+
+        # 1. Encoder
+        hs_pad, hlens, _ = self.enc(hs_pad, hlens)
+
+        # 2. Decoder
+        hlens = torch.tensor(list(map(int, hlens)))
+
+        if recog_args.search_type == 'greedy':
+            y = self.dec.recognize_batch(hs_pad, hlens, recog_args)
+        else:
+            y = self.dec.recognize_beam_batch(hs_pad, hlens, recog_args, rnnlm)
+
+        if prev:
+            self.train()
+
+        return y
+
+    def enhance(self, xs):
+        """Forwarding only the frontend stage
+
+        Args:
+            xs (ndarray): input acoustic feature (T, C, F)
+
+        Returns:
+            enhanced (ndarray):
+            mask (torch.Tensor):
+            ilens (torch.Tensor): batch of lengths of input sequences (B)
+        """
+
+        if self.frontend is None:
+            raise RuntimeError('Frontend does\'t exist')
+        prev = self.training
+        self.eval()
+        ilens = np.fromiter((xx.shape[0] for xx in xs), dtype=np.int64)
+
+        # subsample frame
+        xs = [xx[::self.subsample[0], :] for xx in xs]
+        xs = [to_device(self, to_torch_tensor(xx).float()) for xx in xs]
+        xs_pad = pad_list(xs, 0.0)
+        enhanced, hlensm, mask = self.frontend(xs_pad, ilens)
+
+        if prev:
+            self.train()
+
+        return enhanced.cpu().numpy(), mask.cpu().numpy(), ilens
+
+    def calculate_all_attentions(self, xs_pad, ilens, ys_pad):
+        """E2E attention calculation
+
+        Args:
+            xs_pad (torch.Tensor): batch of padded input sequences (B, Tmax, idim)
+            ilens (torch.Tensor): batch of lengths of input sequences (B)
+            ys_pad (torch.Tensor): batch of padded character id sequence tensor (B, Lmax)
+
+        Returns:
+            att_ws (ndarray): attention weights with the following shape,
+                1) multi-head case => attention weights (B, H, Lmax, Tmax),
+                2) other case => attention weights (B, Lmax, Tmax).
+        """
+
+        with torch.no_grad():
+            # 0. Frontend
+            if self.frontend is not None:
+                hs_pad, hlens, mask = self.frontend(to_torch_tensor(xs_pad), ilens)
+                hs_pad, hlens = self.feature_transform(hs_pad, hlens)
+            else:
+                hs_pad, hlens = xs_pad, ilens
+
+            # encoder
+            hpad, hlens, _ = self.enc(hs_pad, hlens)
+
+            # decoder
+            if self.rnnt_mode == 1:
+                att_ws = self.dec.calculate_all_attentions(hpad, hlens, ys_pad)
+            else:
+                att_ws = []
```

Comment:
I think you can `return []` at the beginning of this method. If so, we can skip the computation in frontend and encoder.